WordPress/Classes/ViewRelated/NUX/LoginEmailViewController.swift

```diff
@@ -1,35 +1,328 @@
+    ///
+    func signinWithUsernamePassword(_ immediateSignin: Bool = false) {
         performSegue(withIdentifier: .showWPComLogin, sender: self)
     }
 
-    override func signinToSelfHostedSite() {
+
+    /// Displays the self-hosted sign in form.
+    ///
+    func signinToSelfHostedSite() {
         performSegue(withIdentifier: .showSelfHostedLogin, sender: self)
     }
+
+
+    /// Proceeds along the "magic link" sign-in flow, showing a form that let's
+    /// the user request a magic link.
+    ///
+    func requestLink() {
+        performSegue(withIdentifier: .startMagicLinkFlow, sender: self)
+    }
+
+
+    /// Validates what is entered in the various form fields and, if valid,
+    /// proceeds with the submit action.
+    ///
+    func validateForm() {
+        displayError(message: "")
+        guard loginFields.username.isValidEmail() else {
+            assertionFailure("Form should not be submitted unless there is a valid looking email entered.")
+            return
+        }
+
+        configureViewLoading(true)
+        let service = AccountService(managedObjectContext: ContextManager.sharedInstance().mainContext)
+        service.isPasswordlessAccount(loginFields.username,
+                                      success: { [weak self] (passwordless: Bool) in
+                                        self?.configureViewLoading(false)
+                                        self?.loginFields.passwordless = passwordless
+                                        self?.requestLink()
+            },
+                                      failure: { [weak self] (error: Error) in
+                                        WPAppAnalytics.track(.loginFailed, error: error)
+                                        DDLogSwift.logError(error.localizedDescription)
+                                        guard let strongSelf = self else {
+                                            return
+                                        }
+                                        strongSelf.configureViewLoading(false)
+
+                                        let userInfo = (error as NSError).userInfo
+                                        if let errorCode = userInfo[WordPressComRestApi.ErrorKeyErrorCode] as? String, errorCode == "unknown_user" {
+                                            let msg = NSLocalizedString("This email address is not registered on WordPress.com.",
+                                                                        comment: "An error message informing the user the email address they entered did not match a WordPress.com account.")
+                                            self?.displayError(message: msg)
+                                        } else {
+                                            strongSelf.displayError(error as NSError, sourceTag: strongSelf.sourceTag)
+                                        }
+        })
+    }
+
+
+    func displayError(message: String) {
```

Comment:
This method should probably have documentation. I was confused why it was being called in the first line of `validateForm()` 😉